### PR TITLE
Allow reloading config without restarting the daemon

### LINKED ISSUE #13: Reload configuration
```
kanshictl reload
```

Requires IPC.

### PR DESCRIPTION
**Pull request by ghost**
Closes #13. There are two supported ways to do it:

1. Run the command `kanshictl reload`
2. Send the process a SIGHUP

The varlink protocol is used, so that should make it easy to add more commands to kanshictl.

### CODE REVIEW

**Comment by ghost** [comment]
This is broken in CI because varlink is not in the AUR. It can be added as a subproject if we want to fix that in the meantime.

**Comment by emersion** [comment]
Added [libvarlink](https://aur.archlinux.org/packages/libvarlink/) to the AUR.

**Comment by ghost** [comment]
Looks good on CI, ready for review

File: include/ipc.h [hunk 1]
```diff
@@ -0,0 +1,28 @@
+#ifndef KANSHI_IPC_H
+#define KANSHI_IPC_H
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <varlink.h>
+
+#define varlink_return_if_fail_v(_val, _fmt, ...) { \
```
Comment (emersion) [inline include/ipc.h]: Please don't use macros.

File: include/kanshi.h [hunk 2]
```diff
@@ -5,6 +5,9 @@
 #include <wayland-client.h>
 
 struct zwlr_output_manager_v1;
+#ifdef KANSHI_HAS_VARLINK
+typedef struct VarlinkService VarlinkService;
```
Comment (emersion) [inline include/kanshi.h]: Why is this necessary?

File: main.c [hunk 3]
```diff
@@ -239,6 +253,8 @@ static void apply_profile(struct kanshi_state *state,
 	}
 
 	zwlr_output_configuration_v1_apply(config);
+
+	wl_display_roundtrip(state->display);
```
Comment (emersion) [inline main.c]: Why is this necessary?

File: main.c [hunk 4]
```diff
@@ -462,22 +482,183 @@ static struct kanshi_config *read_config(void) {
...
+static int do_poll(struct pollfd *fds, nfds_t nfds) {
+	int ret;
+	do {
+		ret = poll(fds, nfds, -1);
+	} while (ret == -1 && errno == EINTR);
+	return ret;
+}
+
+enum readfds_type {
+	FD_WAYLAND		= 0,
```
Comment (emersion) [inline main.c]: Style: don't align

File: main.c [hunk 5]
```diff
@@ -462,22 +482,183 @@ static struct kanshi_config *read_config(void) {
...
+
+	return EXIT_SUCCESS;
+
+read_error:
+	wl_display_cancel_read(state->display);
+	return EXIT_FAILURE;
+}
+
+#ifdef KANSHI_HAS_VARLINK
+long fr_emersion_kanshi_Reload(VarlinkService *service, VarlinkCall *call,
```
Comment (emersion) [inline main.c]: This can be `static`. Can we rename it to `handle_reload`?

File: main.c [hunk 6]
```diff
@@ -462,22 +482,183 @@ static struct kanshi_config *read_config(void) {
...
+		varlink_service_free(service);
+		return NULL;
+	}
+
+	return service;
+}
+#endif
+
 int main(int argc, char *argv[]) {
+	struct wl_display *display = NULL;
```
Comment (emersion) [inline main.c]: Why is moving this variable necessary?

File: main.c [hunk 7]
```diff
@@ -488,12 +669,21 @@ int main(int argc, char *argv[]) {
...
 		fprintf(stderr, "compositor doesn't support "
 			"wlr-output-management-unstable-v1\n");
-		return EXIT_FAILURE;
+		goto error;
 	}
 
-	while (state.running && wl_display_dispatch(display) != -1) {
-		// This space intentionally left blank
+	int ret = loop(&state);
+	goto done;
```
Comment (emersion) [inline main.c]: This goto-logic is complicated. Can we simplify it?

File: main.c [hunk 8]
```diff
@@ -462,22 +482,183 @@ static struct kanshi_config *read_config(void) {
...
+		destroy_config(state->config);
+		state->config = config;
+		state->pending_profile = NULL;
+		state->current_profile = NULL;
+		return try_apply_profiles(state);
+	}
+	return false;
+}
+
+static int do_poll(struct pollfd *fds, nfds_t nfds) {
```
Comment (emersion) [inline main.c]: Maybe we can extract all event loop functions in e.g. `event-loop.c`?

File: main.c [hunk 9]
```diff
@@ -462,22 +482,183 @@ static struct kanshi_config *read_config(void) {
...
+	}
+
+	return EXIT_SUCCESS;
+
+read_error:
+	wl_display_cancel_read(state->display);
+	return EXIT_FAILURE;
+}
+
+#ifdef KANSHI_HAS_VARLINK
```
Comment (emersion) [inline main.c:535]: Instead of having this big `#ifdef`, maybe we can move this into a separate file?

File: main.c [hunk 10]
```diff
@@ -462,22 +482,183 @@ static struct kanshi_config *read_config(void) {
...
+	readfds[FD_VARLINK].fd = varlink_service_get_fd(state->service);
+	readfds[FD_VARLINK].events = POLLIN;
+#endif
+
+	struct pollfd writefds[1];
+	writefds[0].fd = readfds[FD_WAYLAND].fd;
+	writefds[0].events = POLLOUT;
+
+	while (state->running) {
+	 while (wl_display_prepare_read(state->display) != 0) {
```
Comment (emersion) [inline main.c]: Style: identation issue here

File: ctl.c [hunk 11]
```diff
@@ -0,0 +1,78 @@
+
```
Comment (emersion) [inline ctl.c]: Style nit: stray newline

File: ctl.c [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+		VarlinkObject *parameters, uint64_t flags, void *userdata) {
+	bool *done = userdata;
+	*done = true;
+	return 0;
+}
+
+#define ARRAY_LENGTH(_arr) (sizeof(_arr) / sizeof(*(_arr)))
+
+int wait_for_event(VarlinkConnection *connection, bool *done) {
+	struct pollfd fds[1];
```
Comment (emersion) [inline ctl.c]: Do we really need `poll` here? This utility is never going to do anything besides calling varlink functions, so I think blocking is fine.

File: kanshictl.1.scd [hunk 13]
```diff
@@ -0,0 +1,28 @@
+kanshi(1)
```
Comment (emersion) [inline kanshictl.1.scd]: Typo: `kanshictl`

File: meson.build [hunk 14]
```diff
@@ -63,6 +63,18 @@ executable(
 	install: true,
 )
 
+if varlink.found()
+	executable(
+			meson.project_name() + 'ctl',
+			files(
+				'ctl.c'
+				),
```
Comment (emersion) [inline meson.build]: Style: remove an indent level here

File: meson.build [hunk 15]
```diff
@@ -63,6 +63,18 @@ executable(
...
+			include_directories: include_directories('include'),
+			dependencies: [varlink],
+			install: true,
+			)
```
Comment (emersion) [inline meson.build]: Style: remove an indent level here

**Review by emersion — CHANGES_REQUESTED** [review]
Here's a first round of comments. Thanks for your work and sorry for the long delay! I'm excited for this new feature :)

If it's fine by you, I'd prefer to merge the varlink support first and then the signal stuff in another review round.

**Comment by zsugabubus** [comment]
Change my mind, but is it _really_ necessary to add such a bloat? Config reload (that you almost never need) and profile switching could be (easily) done in other ways too.

> The varlink protocol is used, so that should make it easy to add more commands to kanshictl.

What other commands you want to add (don't misunderstand!) to a sh*tty program which only task is to load preset configurations?

**Comment by ghost** [comment]
@emersion Sorry for the delay, no time to work on Wayland stuff recently, but I will be getting to this soon.

@zsugabubus The other use case for IPC is documented in #10. The original implementation used a simple custom IPC, but that was not flexible and was going to be even more code to integrate it well. My personal use case for this is in cyclopsian/wdisplays#4.

**Comment by zsugabubus** [comment]
Why not just kill and restart?

**Comment by ghost** [comment]
Without an IPC mechanism there is no way to find the right process. Doing `killall kanshi` is not safe.

**Comment by ghost** [comment]
It should also be noted that in general, procfs + Unix signals are not appropriate for IPC. See these articles for some more info on the problem and why either a workaround or a kernel-level solution is needed: https://lwn.net/Articles/773459/ https://lwn.net/Articles/794707/

**Comment by zsugabubus** [comment]
> Doing killall kanshi is not safe.

I don't get it. What do you mean by unsafe? Even if you have parallel Wayland session with the same user, you could use:
* PID files,
* add a simple signal handler,
* wrap it inside a (systemd) service.

They would just about 10-15 LOC instead of adding a noname library(*) as dependency plus more hundred lines. For a task that you do only once... okay... twice in your lifetime.

*: Dunno if anybody have tried to build it at all but meson fails with an error.

> It should also be noted that in general, procfs + Unix signals are not appropriate for IPC. See these articles for some more info on the problem [...]

Do you think it is really relevant in our case?

**Comment by ghost** [comment]
The original implementation did that and there were issues so I scrapped it. A daemon should never ever manage its own PID file. If the program is killed then the PID file will stick around and could potentially point to the wrong process. We could only use systemd if it was running in user mode, which not everybody does. Also systemd is optional in wlroots/sway so we want to retain that. Even with systemd there is still no reasonable way to use unix signals for IPC and have it work in all cases, read those articles for more information about this. 

I encourage you to read more about varlink, it supports a very simple process-to-process IPC. Using pid files and signals is not a realistic alternative for IPC. The only realistic alternative is dbus which I would like to avoid if we can because that would be unnecessary; we don't need another daemon to handle multicast. If you're having a problem with building varlink, please confirm that it is a bug in the package and not a missing meson dependency. Then file an issue with varlink if appropriate.

**Comment by zsugabubus** [comment]
###### I would like to stay constructive.

Of course, everything is true what you write. I just saw a simple task, but the solution is an external library (installing, updating, what if abandoned), JSON and more thousands(!) LOC.

When you want to change your background you restart `swaybg`; when you want to update your `sway` config you restart it. For me it would be totally okay if `kanshi` would work the same (although it's not only for `sway`).

In addition, I have _never_ met a program that had a special ability for rereading the configuration. (Maybe not without reason?)

> If the program is killed then the PID file will stick around and could potentially point to the wrong process

Of course. Add a lockfile. ~With this, it's way more race-condition free than varlink is now.~

> I encourage you to read more about varlink, it supports a very simple process-to-process IPC. 

It's surely a nice thing for a process that indent to handle more than one, parameterless command. (Switching doesn't need to be an IPC command. See my PR how I imagined it could work.)

> The only realistic alternative is d[...]

Don't continue. 😄 

> Even with systemd there is still no reasonable way to use unix signals for IPC and have it work in all cases

In my service-like (systemd...) version you don't need signals anymore. You just restart the right instance of the service to reload the config.

---

It would be good to hear what @emersion thinks about it. But I guess he is okay with it, otherwise he wouldn't add it to AUR...

I think I stop trying convincing you. 😭

**Comment by ghost** [comment]
Sway is a bad example. It also includes JSON and many thousand of lines of code to implement the i3 IPC protocol over a socket. When you type `swaymsg reload` that does not kill the program, it sends an IPC message to re-read the config. Your service manager can track PIDs of its child processes itself which is why it's safe to use it to send signals, although this is not totally race-free unless it uses pidfd or something similar.

A lock file is adequate while the process is running, but still has race conditions in this use case. For example, if the program is killed after the other program tries to read the lock file. File locking is also unfortunately not an adequate primitive for IPC.

**Comment by ghost** [comment]
BTW, systemd is now using varlink for the homed API: https://systemd.io/USER_GROUP_API/

**Comment by emersion** [comment]
I agree with @cyclopsian. Adding a simple IPC mechanism doesn't add a lot of complexity.

File: include/ipc.h (hunk 1 above)
Comment (ghost) [reply include/ipc.h]: These macros have been removed.

File: include/kanshi.h (hunk 2 above)
Comment (ghost) [reply include/kanshi.h]: Took this out, it's not necessary and it was only for consistency with the typedef that varlink uses.

File: main.c (hunk 6 above)
Comment (ghost) [reply main.c]: Can't remember, moved it back.

File: main.c (hunk 8 above)
Comment (ghost) [reply main.c]: Moved into a new file.

File: main.c (hunk 9 above)
Comment (ghost) [reply main.c:535]: Moved into ipc.c.

File: ctl.c (hunk 12 above)
Comment (ghost) [reply ctl.c]: We don't, I managed to remove it by unsetting O_NONBLOCK.

File: main.c (hunk 7 above)
Comment (ghost) [reply main.c]: Removed the `error` label, now there is only `done`.

File: main.c (hunk 3 above)
Comment (ghost) [reply main.c]: Okay so this is a bit confusing why it needed to be in this specific spot, but the roundtrip is because the response to the reload IPC message is not really useful unless it's sent after the new output configuration gets applied. Without that there is a race condition. Demonstrating with a shell script that detects new monitors and then tries to automatically configure them and set a background or something like that:

```bash
outputs=$(swaymsg -rt get_outputs)
# ... process the outputs and detect new ones, generate a config ...
echo $myconfig >> ~/.config/kanshi/config
kanshictl reload
swaybg $bgconfig
```

If `kanshictl reload` returns immediately without waiting for this roundtrip, then swaybg could potentially run before the resolution changes. Note that the "just do one roundtrip" method works in sway (and wayfire I think) because it applies the output changes immediately. But I suppose other compositors could wait a frame, or be multi-threaded, or pop up a dialog to confirm the changes first... all of which would also cause this type of script to fail. The workaround in that case (or just without the roundtrip) would be to repeatedly call `swaymsg -t get_outputs` in a while loop after kanshictl, to confirm the resolution changes, which is not great. So I just threw the roundtrip in.

**Comment by ghost** [comment]
This is updated, can squash/reorganize commits when everything is ready.

File: main.c (hunk 3 above)
Comment (ghost) [reply main.c]: Actually it just occurred to me that this can be replaced with a wl_display_sync in the varlink callback.

File: README.md [hunk 16]
```diff
@@ -15,6 +15,7 @@ Dependencies:
 
 * wayland-client
 * scdoc (optional, for man pages)
+* libvarlink
```
Comment (emersion) [inline README.md:18]: Would be nice to explain that it's optional.

File: main.c [hunk 17]
```diff
@@ -479,6 +496,47 @@ static struct kanshi_config *read_config(void) {
...
+		state->current_profile = NULL;
+		return try_apply_profiles(state);
+	}
+	return false;
+}
+
+int kanshi_main_loop(struct kanshi_state *state);
+
+#ifdef KANSHI_HAS_VARLINK
+int kanshi_init_ipc(struct kanshi_state *state);
```
Comment (emersion) [inline main.c:536]: Can we put these decls in a header file `ipc.h`?

File: include/ipc.h [hunk 18]
```diff
@@ -0,0 +1,15 @@
+#ifndef KANSHI_IPC_H
+#define KANSHI_IPC_H
+
+#include <stdio.h>
+#include <stdlib.h>
+
+static int get_ipc_address(char *address, size_t size) {
```
Comment (emersion) [inline include/ipc.h:7]: We don't put code in headers

File: event-loop.c [hunk 19]
```diff
@@ -0,0 +1,186 @@
...
+		return -1;
+	}
+	return 0;
+}
+
+static int signal_pipefds[2];
+
+static void signal_handler(int signum) {
+	if (write(signal_pipefds[1], &signum, sizeof(signum)) == -1) {
+		_exit(signum | 0x80);
```
Comment (emersion) [inline event-loop.c:55]: Maybe print an error message to stderr too?

File: ipc.c [hunk 20]
```diff
@@ -0,0 +1,64 @@
...
+static struct wl_callback_listener reload_config_listener = {
+	.done = reload_config_done
+};
+
+static long handle_reload(VarlinkService *service, VarlinkCall *call,
+		VarlinkObject *parameters, uint64_t flags, void *userdata) {
+	struct kanshi_state *state = userdata;
+	kanshi_reload_config(state);
+	// ensure that changes are applied on the server before sending the response
+	struct wl_callback *callback = wl_display_sync(state->display);
```
Comment (emersion) [inline ipc.c:24]: Ah, I see. This only ensures that the server has received the configuration request, the server is free to wait an arbitrary amount of time before applying the configuration.

We can add a TODO about using the wlr-output-management event instead.

File: ipc.c [hunk 21]
```diff
@@ -0,0 +1,64 @@
...
+	wl_callback_add_listener(callback, &reload_config_listener, call);
+	return 0;
+}
+
+int kanshi_init_ipc(struct kanshi_state *state) {
+	VarlinkService *service;
+	char address[PATH_MAX];
+	get_ipc_address(address, sizeof(address));
+	if (varlink_service_new(&service,
+			"emersion", "kanshi", "1.1", "https://wayland.emersion.fr/kanshi/",
```
Comment (emersion) [inline ipc.c:34]: Can we get this version number from a `-D` flag set by Meson?

**Review by emersion — COMMENTED** [review]
On the whole this looks pretty good to me! After these comments are addressed and the commits are re-organized I'll merge this.

Thanks for your work!

**Comment by somini** [comment]
What's the status of this?

This blocks https://github.com/cyclopsian/wdisplays/issues/4, from what I can tell.

**Comment by ghost** [comment]
Planning to finish this soon, due to current events I've been away from the main multi-monitor setup I have but I think I can find another way to test it.

**Comment by ghost** [comment]
Now that kanshi is in debian unstable, I'm trying to get libvarlink into debian before I finish this up https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=960434

**Comment by maricn** [comment]
Just trying to get a sense if I should look into checking this branch out locally or if there's any chance @cyclopsian would be picking up the PR any time soon? And thanks for the work so far!

**Comment by maricn** [comment]
> Just trying to get a sense if I should look into checking this branch out locally or if there's any chance @cyclopsian would be picking up the PR any time soon? And thanks for the work so far!

OK, so this user has left GitHub, and is now replaced with the ghost account. 👻 

I'm able to checkout this branch at `varlink`, but I don't code in C and it will take me another year to push this through. As the OP has left the GitHub, this is basically an open call for anyone to jump in, merge this with the mainline branch and address the rather minor PR comments, so - any takers?

**Comment by emersion** [comment]
Yes, anyone's welcome to jump it and adopt this work, keeping the attribution if applicable. I'd prefer to have two separate pull requests though (one for signal handling, another one for IPC).

**Comment by ericonr** [comment]
Assuming no one's started work on this yet, I will try my hand on finishing up the PR.

**Comment by emersion** [comment]
Superseded by https://github.com/emersion/kanshi/pull/107
